### PR TITLE
added version check for CLI binary

### PR DESCRIPTION
**Pull request by jasmingacic**
# Summary of this proposal

Checking CLI version before running subcommands .

```
# ./calyptia get core_instances 
Warning: This version 1.0.0 of Calyptia cli is outdated. The latest version available is 1.12.1
NAME        VERSION ENVIRONMENT PIPELINES TAGS STATUS      AGE
testsecrets v2.1.1  default     2              unreachable 1 hour
```
disabling the check
```
# ./calyptia get core_instances --disable-version-check
NAME        VERSION ENVIRONMENT PIPELINES TAGS STATUS      AGE
testsecrets v2.1.1  default     2              unreachable 1 hour
```
Check is disabled by checking **isatty** so when piping the output won't be messed up

```
# ./calyptia get core_instances -o yaml | yq
- id: 4d74f2cb-a3ba-4cec-8d16-a9c5f8b95ffb
  token: ....
```

## Asana task(s) link.

https://app.asana.com/0/1204316201844829/1206101751551634

## Notes for the reviewer

Ideally we should be doing checks for image update but I would rather rework update to do reinstall of the operator to ensure that the whole package is indeed picking up the latest version.

### CODE REVIEW

File: cmd/root.go [hunk 1]
```diff
@@ -85,13 +90,19 @@ func NewRootCmd(ctx context.Context) *cobra.Command {
...
+			}
+		},
 	}
 
 	cmd.SetOut(os.Stdout)
 
 	fs := cmd.PersistentFlags()
 	fs.StringVar(&cloudURLStr, "cloud-url", cfg.Env("CALYPTIA_CLOUD_URL", cloudURLStr), "Calyptia Cloud URL")
 	fs.StringVar(&token, "token", cfg.Env("CALYPTIA_CLOUD_TOKEN", token), "Calyptia Cloud Project token")
+	fs.BoolVar(&vercheck, "disable-version-check", false, "disable version check ")
```
Comment (patrick-stephens) [inline cmd/root.go:105]: Let's add an env var as well so we don't have to keep adding the flag or set up an alias.

File: cmd/version/version.go [hunk 2]
```diff
@@ -6,7 +6,7 @@ import (
 
 var (
 	DefaultCloudURLStr = "https://cloud-api.calyptia.com"
-	Version            = "dev" // To be injected at build time:  -ldflags="-X 'github.com/calyptia/cli/cmd/version.Version=xxx'"
```
Comment (patrick-stephens) [inline cmd/version/version.go:9]: How come you changed this?

File: cmd/version/githubclient.go [hunk 3]
```diff
@@ -0,0 +1,219 @@
+package version
```
Comment (patrick-stephens) [inline cmd/version/githubclient.go:1]: Why not use an existing library like https://github.com/google/go-github?
Feels like we're reinventing stuff and have to then maintain it.

File: cmd/root.go [hunk 4]
```diff
@@ -104,8 +115,40 @@ func NewRootCmd(ctx context.Context) *cobra.Command {
...
+		cliversion.NewVersionCommand(),
 	)
 
 	return cmd
 }
+
+func versionCheck(cmd *cobra.Command) {
+	if isatty.IsTerminal(os.Stdout.Fd()) {
+		if len(cliversion.Version) != 0 && cliversion.Version != "dev" {
+			ghclient, err := utils.NewGithubClient("", nil)
```
Comment (patrick-stephens) [inline cmd/root.go:127]: Rate limiting will presumably just drop out silently along with any other http errors?

File: cmd/version/version.go (hunk 2 above)
Comment (jasmingacic) [reply cmd/version/version.go:9]: forgot to return it while testing

File: cmd/version/githubclient.go (hunk 3 above)
Comment (jasmingacic) [reply cmd/version/githubclient.go:1]: go-github has lot more than what we need. So I just took that one part without importing another library.

File: cmd/root.go (hunk 4 above)
Comment (jasmingacic) [reply cmd/root.go:127]: If it kicks in it will be ignored

File: cmd/root.go (hunk 1 above)
Comment (jasmingacic) [reply cmd/root.go:108]: done

Comment (niedbalski) [reply cmd/root.go:108]: +1 to this

Comment (jasmingacic) [reply cmd/root.go:108]: https://github.com/calyptia/cli/pull/714/files#diff-ab967ab1a2f3a1b769106eeb7bfe892ef0e81d1d27811fa15be08e6749feee1fR31

File: cmd/root.go [hunk 5]
```diff
@@ -104,8 +116,40 @@ func NewRootCmd(ctx context.Context) *cobra.Command {
...
 		newCmdDelete(config),
 		newCmdWatch(config),
-		version.NewVersionCommand(),
+		cliversion.NewVersionCommand(),
 	)
 
 	return cmd
 }
+
+func versionCheck(cmd *cobra.Command) {
```
Comment (nicolasparada) [inline cmd/root.go:125]: There are two `if`s on this function that should be inverted to reduce indentation.

File: cmd/root.go [hunk 6]
```diff
@@ -9,15 +9,20 @@ import (
...
 
+	"github.com/hashicorp/go-version"
+	"github.com/mattn/go-isatty"
 	"github.com/spf13/cobra"
 
 	cloudclient "github.com/calyptia/api/client"
 	cnfg "github.com/calyptia/cli/cmd/config"
-	"github.com/calyptia/cli/cmd/version"
+	cliversion "github.com/calyptia/cli/cmd/version"
+	utils "github.com/calyptia/cli/cmd/version"
```
Comment (nicolasparada) [inline cmd/root.go:19]: `utils` is a duplicate import of `cliversion` (An also a terrible name).

File: cmd/version/version.go [hunk 7]
```diff
@@ -6,7 +6,7 @@ import (
...
+	Version            = "1.0.0" // To be injected at build time:  -ldflags="-X 'github.com/calyptia/cli/cmd/version.Version=xxx'"
```
Comment (nicolasparada) [inline cmd/version/version.go:9]:
```suggestion
	Version            = "dev" // To be injected at build time:  -ldflags="-X 'github.com/calyptia/cli/cmd/version.Version=xxx'"
```

File: cmd/version/githubclient.go [hunk 8]
```diff
@@ -0,0 +1,219 @@
...
+	rel, err := url.Parse(path)
+	if err != nil {
+		return nil, err
+	}
+
+	u := c.baseURL.ResolveReference(rel)
+	var req *http.Request
+	if body != nil {
+		bodyBytes, _ := json.Marshal(body)
+		req, _ = http.NewRequest(method, u.String(), bytes.NewBuffer(bodyBytes))
```
Comment (nicolasparada) [inline cmd/version/githubclient.go:89]: missing error handling

File: cmd/version/githubclient.go [hunk 9]
```diff
@@ -0,0 +1,219 @@
...
+		req.Header.Set("Content-Type", "application/json")
+	} else {
+		req, _ = http.NewRequest(method, u.String(), nil)
```
Comment (nicolasparada) [inline cmd/version/githubclient.go:92]: missing error handling

File: cmd/version/githubclient.go [hunk 10]
```diff
@@ -0,0 +1,219 @@
...
+	}
+	rel, err := url.Parse(path)
+	if err != nil {
+		return nil, err
+	}
+
+	u := c.baseURL.ResolveReference(rel)
+	var req *http.Request
+	if body != nil {
+		bodyBytes, _ := json.Marshal(body)
```
Comment (nicolasparada) [inline cmd/version/githubclient.go:88]: missing error handling

File: cmd/version/githubclient.go [hunk 11]
```diff
@@ -0,0 +1,219 @@
...
+
+func (c *GithubClient) Do(req *http.Request, v interface{}) (*ErrorResponse, error) {
+	resp, err := c.client.Do(req)
+	if err != nil {
+		return nil, err
+	}
+
+	defer resp.Body.Close()
+	if resp.StatusCode > 299 {
+		o, _ := io.ReadAll(resp.Body)
```
Comment (nicolasparada) [inline cmd/version/githubclient.go:112]: missing error handling

File: cmd/version/githubclient.go [hunk 12]
```diff
@@ -0,0 +1,219 @@
...
+		errResp := &ErrorResponse{
+			StatusCode: resp.StatusCode,
+			Errors:     string(o),
+		}
+
+		return errResp, fmt.Errorf("%s returned %d", req.URL, resp.StatusCode)
```
Comment (nicolasparada) [inline cmd/version/githubclient.go:118]: I don't get this API... This returns an `error` but also a response...
`*ErrorResponse` should instead implement `error` and return just that. Not both values.

File: cmd/version/githubclient.go [hunk 13]
```diff
@@ -0,0 +1,219 @@
...
+	return client, nil
+}
+
+type GithubClient struct {
+	client  *http.Client
+	apiKey  string
+	baseURL *url.URL
+}
+
+type ErrorResponse struct {
```
Comment (nicolasparada) [inline cmd/version/githubclient.go:61]:
```suggestion
type GitHubError struct {
```

File: cmd/version/githubclient.go [hunk 14]
```diff
@@ -0,0 +1,219 @@
...
+	"encoding/json"
+	"fmt"
+	"io"
+	"net/http"
+	"net/url"
+	"time"
+)
+
+const (
+	baseURL = "https://api.github.com/repos/calyptia/"
```
Comment (nicolasparada) [inline cmd/version/githubclient.go:14]:
```suggestion
	githubBaseURL = "https://api.github.com/repos/calyptia/"
```

File: cmd/version/githubclient.go [hunk 15]
```diff
@@ -0,0 +1,219 @@
...
+
+	err = json.Unmarshal(o, v)
+	if err != nil {
+		return nil, err
+	}
+
+	return nil, err
+}
+
+type Tag struct {
```
Comment (nicolasparada) [inline cmd/version/githubclient.go:134]: All these types are for the github API client, but are on the `version` package...
Maybe rename them to include the work "github" or move the entire github client to a new package.

File: cmd/version/githubclient.go [hunk 16]
```diff
@@ -0,0 +1,219 @@
...
+		SubscriptionsURL  string `json:"subscriptions_url"`
+		OrganizationsURL  string `json:"organizations_url"`
+		ReposURL          string `json:"repos_url"`
+		EventsURL         string `json:"events_url"`
+		ReceivedEventsURL string `json:"received_events_url"`
+		Type              string `json:"type"`
+		SiteAdmin         bool   `json:"site_admin"`
+	} `json:"author"`
+	NodeID          string    `json:"node_id"`
+	TagName         string    `json:"tag_name"`
```
Comment (nicolasparada) [inline cmd/version/githubclient.go:174]: Only `tag_name` is used really. So there is no need to unmarshall and keep track of all the fields.

File: cmd/version/githubclient.go [hunk 17]
```diff
@@ -0,0 +1,219 @@
...
+
+	path := fmt.Sprintf("%s/releases/latest", repo)
+	if resp, err := c.DoRequest("GET", path, nil, i); err == nil {
+		return i, resp, err
+	} else {
+		return nil, resp, err
+	}
+}
+
+func (c *GithubClient) GetLatest(repo string) (string, error) {
```
Comment (nicolasparada) [inline cmd/version/githubclient.go:28]:
```suggestion
func (c *GithubClient) GetLatestTag(repo string) (string, error) {
```

**Review by nicolasparada — CHANGES_REQUESTED** [review]
Left a few comments.

File: cmd/version/githubclient.go [hunk 18]
```diff
@@ -0,0 +1,219 @@
...
+	i, _, err := c.GetTags(repo)
+	if err != nil {
+		return "", err
+	}
+
+	return i.TagName, nil
+}
+
+func NewGithubClient(apiKey string, httpClient *http.Client) (*GithubClient, error) {
```
Comment (nicolasparada) [inline cmd/version/githubclient.go:37]: I don't think constructors should error.

File: cmd/version/githubclient.go (hunk 8 above)
Comment (jasmingacic) [reply cmd/version/githubclient.go:89]: it is intentional

File: cmd/version/githubclient.go (hunk 12 above)
Comment (jasmingacic) [reply cmd/version/githubclient.go:118]: there is no need for this actually

Comment (jasmingacic) [reply cmd/version/githubclient.go:118]: all errors are ignored here anyways

File: cmd/version/githubclient.go (hunk 16 above)
Comment (jasmingacic) [reply cmd/version/githubclient.go]: done

**Review by niedbalski — APPROVED** [review]
LGTM

**Review by patrick-stephens — APPROVED** [review]

**Comment by patrick-stephens** [comment]
@nicolasparada please approve

**Comment by niedbalski** [comment]
> I'll approve for now but I don't like this approach of ignoring all the errors and just continuing... It's not the way of how Go works.

@nicolasparada let's don't merge this if we have no agreement. Can you expand on your concerns here?

**Comment by jasmingacic** [comment]
This has been sitting here since December 
I'm closing it as it has become stale

**Pull request closed without merging**